infra/blueprint-test/pkg/kpt/setters.go

```diff
@@ -0,0 +1,97 @@
+package kpt
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/GoogleContainerTools/kpt-functions-catalog/functions/go/list-setters/listsetters"
+	"sigs.k8s.io/kustomize/kyaml/kio"
+	"sigs.k8s.io/kustomize/kyaml/kio/kioutil"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+// UpsertSetters inserts or updates setters if apply-setters fn config is discovered.
+func UpsertSetters(nodes []*yaml.RNode, setters map[string]string) error {
+	kf, err := listsetters.FindKptfile(nodes)
```

Comment:
Does this mean we only act on the first Kptfile found (presumably the root)?

Reply:
Correct, `FindKptfile` looks for a PathAnnotation that is `Kptfile` which will only match the root kptfile and not any other Kptfiles. My thought was as `render` happens bottoms up, upserting a setter at the root will affect nested pkgs that have the same setters too. In the future if we want to have a more targeted upsert for nested pkgs we can add that as a separate helper.

Reply:
Sounds right, I think we only need to act on the root.